pages/shared/syncHelpers.js

```diff
@@ -303,12 +310,164 @@ module.exports.gitStatus = function (locals, callback) {
+      });
+    } else {
+      job.info('Repository found');
+      // Already exists, nothing to do
+      callback(null);
+    }
+  });
+}
+
+function logProgressOutput(output, job, printedInfos, prefix) {
+  let info = null;
+  if (
+    'status' in output &&
+    'id' in output &&
+    'progressDetail' in output &&
+    output.progressDetail.total
+  ) {
+    info = `${output.status} ${output.id} (${output.progressDetail.total} bytes)`;
+  } else if ('status' in output && 'id' in output) {
+    info = `${output.status} ${output.id}`;
+  } else if ('status' in output) {
+    info = `${output.status}`;
+  }
+  if (info != null && !printedInfos.has(info)) {
+    printedInfos.add(info);
+    job.info(prefix + info);
+  }
+}
+
+function pullAndPushToECR(image, dockerAuth, job, callback) {
+  debug(`pullAndPushtoECR for ${image}`);
+
+  if (!config.cacheImageRegistry) {
+    return callback(new Error('cacheImageRegistry not defined'));
+  }
+
+  const repository = new DockerName(image);
+  const params = {
+    fromImage: repository.getRepository(),
+    tag: repository.getTag() || 'latest',
+  };
+  job.info(`Pulling ${repository.getCombined()}`);
+  docker.createImage({}, params, (err, stream) => {
+    if (ERR(err, callback)) return;
+
+    const printedInfos = new Set();
+    docker.modem.followProgress(
+      stream,
+      (err) => {
+        if (ERR(err, callback)) return;
+
+        job.info('Pull complete');
+
+        // Find the image we just downloaded
+        const downloadedImage = repository.getCombined(true);
+        job.info(`Locating downloaded image: ${downloadedImage}`);
+        locateImage(downloadedImage, (err, localImage) => {
+          if (ERR(err, callback)) return;
+          job.info('Successfully located downloaded image');
+
+          // Tag the image to add the new registry
+          repository.setRegistry(config.cacheImageRegistry);
+
+          var options = {
+            repo: repository.getCombined(),
+          };
+          job.info(`Tagging image: ${options.repo}`);
+          localImage.tag(options, (err) => {
+            if (ERR(err, callback)) return;
+            job.info('Successfully tagged image');
+
+            const repositoryName = repository.getRepository();
+            job.info(`Ensuring repository exists: ${repositoryName}`);
+            confirmOrCreateECRRepo(repositoryName, job, (err) => {
+              if (ERR(err, callback)) return;
+              job.info('Successfully ensured repository exists');
+
+              // Create a new docker image instance with the new registry name
+              // localImage isn't specific enough to the ECR repo
+              const pushImageName = repository.getCombined();
+              var pushImage = new Docker.Image(docker.modem, pushImageName);
+
+              job.info(`Pushing image: ${repository.getCombined()}`);
+              pushImage.push(
+                {
+                  authconfig: dockerAuth,
+                },
```

Comment:
The only change I made to this copy/pasted code was moving the auth information from the third argument to this `authconfig` property, which is actually reflected in the TypeScript types.